Enable External Media Source Mode (#12)
verbit-ai/verbit-streaming-python-sdk: README.md

    
client = WebSocketStreamingClient(access_token="ACCESS TOKEN")

> response_generator = client.start_with_media(media_generator=media_generator,

Reviewer: @EliAsor We cannot break the existing interface.
The "normal" mode of operation where an audio generator is provided must remain with its current name `start_stream`.
We only need a name for the new method
I think  something like your name or `start_with_external_media` is good - clear and explicit

Reviewer 2: I agree with @roman-verbit-ai , no need to change the external API, the private API currently called `_start_stream()` could be change to `_start_connection()`, or anything that would not include the word `stream` as it now covers the case of a `None Stream`.

Author: @roman-verbit-ai @kobicalev I have no problem with NOT adding the new `start_with_media` method, but I didn't break anything in the existing method, only added a new one, with a more explicit name (as we discussed, if I understood you correctly).
Anyway, I don't mind removing the new method (`start_with_media`) and just keep the `start_stream` method.

Regarding renaming the internal `_start_stream` - I agree that it's better to rename it to something different enough from `start_stream`. The name `_start_connection` is a good option, but maybe `_connect_and_start` is a bit better :) 
WDYT?

Author: @roman-verbit-ai I added a "compatibility note" on the README.md file regarding the deprecation of `start_stream` (while still keeping it for backward-compatibility). Please have a look. If you still feel that we need to keep just `start_stream` and remove `start_with_media` - I'll remove the new method, but I think it's better to have the explicit method names introduced in this version.

Reviewer: @EliAsor I personally think it's confusing to have two functions that do the same thing.
Also, I think the default, I even dare say "normal", way to work with this SDK is the bi-directional WS method.
Therefore, I think it's good that the method is named simply `start_stream`
Only the other, "abnormal" method with external media, should be the one that is named with a special name.

Reviewer 2: @EliAsor: I agree with @roman-verbit-ai, as an API-user, I would find it mentally simpler to have one method here, as they do do the same thing basically, this could also simplify client-code in the general case.
I recall from a f2f discussion about this, that you wanted to make sure client-code does not start a media-less stream by mistake, expecting replies that never arrive.

I would thus embrace the `PEP20` **Explicit is better than implicit**, while keeping the old(er) interface the following way: (indeed no need to start deprecating calls so early down the (pre)release path)

1. Keeping solely `start_stream()`
2. Making an explicit argument about the 'abnormal' `start_with_external_source` case, where the default is `No`, this might be a `bool` or a more general `dict`, TBD (in this PR, and soonish =)
3. None compatible input arguments should `raise` a `ValueError`: E.G.: `start_with_external_source=True` but `media_generator is not None` ; or vice versa
4. A reasonable way to *document* this for users, _could_ use `@typing.overload`
since conceptually this is kind of 'function overloading':
 https://peps.python.org/pep-0484/#function-method-overloading : [ see: 
https://mypy.readthedocs.io/en/stable/more_types.html#function-overloading , it's slightly clearer]

Note that `4` is just a way to *document* the expected combinations as function signature, there's no 'out of the box' way to do function overloading in run-time.

Author: @roman-verbit-ai ok, I removed all references to `start_with_media` and kept only `start_stream`

                                             media_config=media_config,
                                             response_types=response_types)
```

### Providing media via an external source

> It is possible to use an external media source to provide media chunks to the Speech Recognition Service.

Reviewer: ```suggestion
It is possible to use an external media source to provide media to the Speech Recognition Service.
```

To do so, you need to specify the relevant input method when booking the session via Verbit's Ordering API.

In such a scenario, you should **not** provide a media generator to the `WebSocketStreamingClient`. 
Connecting the `WebSocketStreamingClient` to the Speech Recognition Service will initiate the session

> and signal the server to start listening to the external media source.

Reviewer: ```suggestion
and signal the server to start consuming media from the external media source.
```
just to be ambiguous enough to accommodate for both "listening" and "pulling-from"

Therefore, **you should only connect the `WebSocketStreamingClient` to the service after the external media source is ready.**

#### Example

The following example connects to the Speech Recognition Service without providing a media generator:

```python
from verbit.streaming_client import WebSocketStreamingClient, ResponseType

response_types = ResponseType.Transcript | ResponseType.Captions
    
client = WebSocketStreamingClient(access_token="ACCESS TOKEN")

response_generator = client.start_with_external_source(response_types=response_types)
```

### Getting responses

> The client's `start_with_media()` and `start_with_external_source()` methods return a generator which can be iterated to fetch the Speech Recognition responses:

Reviewer: ```suggestion
The client's `start_stream()` and `start_with_external_source()` methods return a generator which can be iterated to fetch the Speech Recognition responses:
```

Author: done.

                                             response_types=response_types)
```

> _Compatibility note: the method `start_stream` is considered deprecated as of version 0.9.0, and is superceded by `start_with_media`, to explicitly denote whether a media generator is to be provided or not. 

Reviewer: As I wrote in another comment, I think `start_stream` is perfectly fine and there's no need for this.
It's not a good enough reason to start messing with deprecation and interface changes... before we even went to real prod :)
Moreover, `start_with_media` sounds strange, I think it's "default enough" that you stream media to an ASR service - so no need to specify it so explicitly.

Author: Removed the compatibility note.

    Example "captions" responses can be found in [examples/responses/captions.md](https://github.com/verbit-ai/verbit-streaming-python-sdk/blob/main/examples/responses/captions.md).

### Error handling and recovery

#### Initial connection

> In case the WebSocket client fails to establish the (initial) connection with the service (e.g. due to temporary unavailability), 

Reviewer: remove parentheses - both instances are important information, not need to demote them using parentheses

> it will perform exponential retry, up to `max_connection_retry_seconds` (configurable).

Reviewer: indicate the name of the property through which this is configurable

Reviewer 2: that's the property name, maybe we can hyperlink this soon to some Swagger gen pages, etc.

Author: Added a link to the property in that place.
